keps/sig-storage/3476-volume-group/README.md

```diff
@@ -0,0 +1,1680 @@
+
+CSI drivers that do not have a volume_group_id concept can use the VolumeGroup name stored in Kubernetes API server as the volume_group_id.
+
+// VolumeGroupSpec describes the common attributes of group storage devices
+// and allows a Source for provider-specific attributes
+Type VolumeGroupSpec struct {
+        // +optional
+        VolumeGroupClassName *string
+
+        // If InitSource is nil, an empty volume group will be created.
+        // Otherwise, a volume group will be created with PVCs.
+        // If Selector is set in InitSource, existing PVCs with matching
+        // label will be added to the volume group.
+        // If SourceVolumeGroupSnapshotName is not nil, the volume group
+        // will be created from the source VolumeGroupSnapshot.
+        // This field determines what PVCs will be in the volume group
+        // when it is initially created. PVCs can be added to or removed
+        // from the volume group later if CSI driver supports
+        // VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME.
+        // +optional
+        InitSource *VolumeGroupSource
+}
+
+// VolumeGroupSource contains 2 options. If VolumeGroupSource is not nil,
+// one and only one of the 2 options must be defined.
+Type VolumeGroupSource struct {
+        // A label query over existing persistent volume claims to be added to the volume group.
+        // +optional
+        Selector *metav1.LabelSelector
+
+        // If specified, the VolumeGroup will be created from the source
+        // VolumeGroupSnapshot.
+        // +optional
+        SourceVolumeGroupSnapshotName *string
+}
+
+
+### Example Yaml Files for Volume Placement
+
+A VolumeGroupClass that supports placement:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupClass
+metadata:
+  name: placementGroupClass1
+spec:
+  parameters:
+     …...
+  allowedTopologies: [failure-domain.example.com/placement: storagePool1]
+```
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroup
+metadata:
+  Name: placemenGroup1
+spec:
+  volumeGroupClassName: placementGroupClass1
+```
+
+A PVC that belongs to both the volume group with groupSnapshot support and placement.
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: pvc1
+  annotations:
+spec:
+  accessModes:
+  - ReadWriteOnce
+  dataSource: null
+  resources:
+        requests:
+        storage: 1Gi
+  storageClassName: storageClass1
+  volumeMode: Filesystem
+  volumeGroupNames: [volumeGroup1, placementGroup1]
+```
+
+If both placement group and volume group with groupSnapshot support are defined, it is possible for the same volume to join both groups. For example, a volume group with groupSnapshot support may include volume members from two placement groups as they belong to the same application.
+
+## Graduation Criteria
+### Alpha
+* Initial feature implementation, including:
+  * Volume group.
+  * Volume group snapshot.
+* Sample implementation in the csi-driver-host-path.
```

Comment:
Are there any real-world CSI drivers that we know can implement this? I think it would be good to have a real storage system to help validate the design.

Reply:
Yes. A few people have reviewed this KEP and confirmed it works for their storage systems.  See details here: 
https://docs.google.com/document/d/1c_48xZuuvyfYrvgIHgwNFf4AHGrbvaUsbRdKfvtRLdQ/edit#heading=h.gnlmzi84mhax